### PR TITLE
Add benchmark project

### PR DESCRIPTION
**Pull request by huoyaoyuan**
The same purpose with ppy/osu-framework#3038 . To benchmark osu! specific stuff.

I can run it with VS and VSCode, but my Rider says it can't build solution. May need investigation.

##### About beatmap and storyboard parsing benchmarks
I choose https://osu.ppy.sh/beatmapsets/261911#osu/736300 as a famous storyboard-full map.
The memory performance is terrible now.

```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
  DefaultJob : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
```

|                    Method |       Mean |     Error |    StdDev |      Gen 0 |     Gen 1 |    Gen 2 | Allocated |
|-------------------------- |-----------:|----------:|----------:|-----------:|----------:|---------:|----------:|
|    BenchmarkLegacyBeatmap |   4.531 ms | 0.0877 ms | 0.0939 ms |   500.0000 |  250.0000 |        - |   3.01 MB |
| BenchmarkLegacyStoryboard | 100.009 ms | 0.5818 ms | 0.4858 ms | 10200.0000 | 2200.0000 | 600.0000 |  57.86 MB |

### CODE REVIEW

**Comment by peppy** [comment]
Will finally give me a place to put by [old benchmark PR](https://github.com/ppy/osu/pull/4523).

File: osu.Game.Benchmarks/osu.Game.Benchmarks.csproj [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="BenchmarkDotNet" Version="0.12.0" />
+    <PackageReference Include="nunit" Version="3.12.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.16.0" />
```
Comment (smoogipoo) [inline osu.Game.Benchmarks/osu.Game.Benchmarks.csproj]:
```suggestion
    <PackageReference Include="NUnit3TestAdapter" Version="3.15.1" />
```

See: https://youtrack.jetbrains.com/issue/RIDER-37769 . Hinders our workflow for now.

File: osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            {
+                var decoder = Decoder.GetDecoder<Beatmap>(reader);
+                return decoder.Decode(reader);
+            }
+        }
+
+        [Benchmark]
+        public Storyboard BenchmarkLegacyStoryboard()
+        {
+            using (var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(typeof(Program), "Resources.MitiS & MaHi - Blu (Speed Up Ver.) (Ashasaki).osb"))
```
Comment (smoogipoo) [inline osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs]: We generally don't include full beatmaps/storyboards for copyright reasons. Would prefer if the ones already included in the game/test resources were used instead. Eventually we would construct beatmaps/storyboards on a per-case basis and drill down these benchmarks.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs]: Maybe worthwhile checking out the way I did it in the PR i mentioned. It generates a beatmap with enough content to be an intensive benchmark without using existing beatmaps.

Comment (huoyaoyuan) [reply osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs]: I'm fine to auto generate beatmaps and storyboards, with several points:
- I need to know specification of legacy beatmap files to generates text that's to be parsed from.
- Patterns coverage.
- A handful entry to run benchmark on ranked maps.

Comment (smoogipoo) [reply osu.Game.Benchmarks/BenchmarkBeatmapParsing.cs]: 1) Create a beatmap and encode it with a `LegacyBeatmapEncoder`. It doesn’t support storyboards right now, though. I’d be fine with omitting it until that’s implemented.
2) Should be made on a per-case basis, like decoding a slider, or decoding a hitcircle, or a hitsound. As above you can just place many of these and encode them to have an “intensive” test.
3) Would love to see something like the beatmap conversion tests, where you can add a new test case based on the file name. That should be enough to then be able to separate out the problematic part and make a custom test for it.

**Comment by huoyaoyuan** [comment]
I tried to write some simple random generation but fails. Keeping it simple now because my purpose was benchmarking other stuff.

Rewritten the history so that the actual beatmap doesn't appear in history. Running on the beatmap bundled in resources instead.

**Review by peppy — APPROVED** [review]
seems fine for now

**Pull request merged**